### PR TITLE
Up to 2.1.0

### PR DESCRIPTION
**Pull request by RaulPPelaez**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

I tried to follow the build instructions in pytorch for this one: https://github.com/pytorch/pytorch#from-source

### CODE REVIEW

**Comment by RaulPPelaez** [comment]
@conda-forge-admin, please rerender

**Comment by RaulPPelaez** [comment]
Some builds fail when testing with:
```shell
import: 'torch'
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698406509295/test_tmp/run_test.py", line 2, in <module>
    import torch
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698406509295/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.8/site-packages/torch/__init__.py", line 234, in <module>
    _load_global_deps()
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698406509295/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.8/site-packages/torch/__init__.py", line 193, in _load_global_deps
    raise err
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698406509295/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.8/site-packages/torch/__init__.py", line 174, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698406509295/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libmkl_intel_lp64.so.2: cannot open shared object file: No such file or directory
```

Others also fail when testing with:
```shell
import: 'torch'
import: 'torch'
+ OMP_NUM_THREADS=4
+ python ./test/run_test.py
/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698406493308/test_tmp/./test/run_test.py:18: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
Ignoring disabled issues:  ['']
Unable to import boto3. Will not be emitting metrics.... Reason: No module named 'boto3'
Missing pip dependency: pytest-rerunfailures, please run `pip install -r .ci/docker/requirements-ci.txt`
+ true
+ pip check
torch 2.1.0 requires fsspec, which is not installed.
Tests failed for pytorch-2.1.0-cpu_generic_py310hefbe716_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for pytorch-2.1.0-cpu_generic_py310hefbe716_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: pytorch-2.1.0-cpu_generic_py310hefbe716_0.conda
##[error]Bash exited with code '1'.
Finishing: Run docker build
```

I added fsspec as a run dep, not sure what to do about the mkl library...

**Comment by hmaarrfk** [comment]
you might need a patch 

:
2023-10-27T18:50:59.4206092Z   /home/conda/feedstock_root/build_artifacts/pytorch-recipe_1698429688496/work/c10/cuda/CUDACachingAllocator.cpp:1144:15: error: 'nvmlProcessInfo_v1_t' was not declared in this scope; did you mean 'nvmlProcessInfo_t'?

**Comment by RaulPPelaez** [comment]
Related https://github.com/pytorch/pytorch/issues/100618

File: recipe/meta.yaml [hunk 1]
```diff
@@ -75,22 +72,30 @@ outputs:
...
+        - types-dataclasses
+        - sympy
+        - filelock
+        - networkx
+        - jinja2
+        - fsspec
+        - packaging
+        - optree
+        - mkl {{ mkl }}   # [blas_impl == "mkl"]
+        - mkl-include      # [blas_impl == "mkl"]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:92]:
```suggestion
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -75,22 +72,30 @@ outputs:
...
+        - hypothesis
+        - numpy
+        - psutil
         - pyyaml
         - requests
-        - future
-        - six
-        - mkl-devel {{ mkl }}   # [blas_impl == "mkl"]
-        - libcblas * *_mkl      # [blas_impl == "mkl"]
+        - setuptools
```
Comment (hmaarrfk) [inline recipe/meta.yaml:82]:
```suggestion
        - mkl-devel {{ mkl }}   # [blas_impl == "mkl"]
        - libcblas * *_mkl      # [blas_impl == "mkl"]
        - setuptools
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -18,11 +18,7 @@ source:
   git_url: https://github.com/pytorch/pytorch.git
   git_tag: v{{ version }}
   patches:
-    - patches/fix_dispatch_apply_auto.patch
-    # 2022/07 hmaarrfk
-    # We've patch out the bug in GCC that makes it incompatible
-    # https://github.com/conda-forge/ctng-compilers-feedstock/blob/7ea10c2526445beb9faff23efcc6bc40fc0079b5/recipe/meta.yaml#L23=
-    - patches/remove_gcc_maximum_requirements.patch
```
Comment (hmaarrfk) [inline recipe/meta.yaml:25]: please remove these patches from the repo if they aren't necessary anymore

File: recipe/meta.yaml [hunk 4]
```diff
@@ -18,11 +18,7 @@ source:
...
+    - nvml.patch  # [cuda_compiler_version == "11.2"]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:21]: please try to follow the old organization, put this file in the `patches/` directory

File: recipe/meta.yaml [hunk 5]
```diff
@@ -75,22 +72,30 @@ outputs:
...
+        - filelock
+        - networkx
+        - jinja2
+        - fsspec
+        - packaging
+        - optree
         - libcblas              # [blas_impl != "mkl"]
         - liblapack             # [blas_impl != "mkl"]
-        - libgomp   # [linux]
-        - llvm-openmp    # [osx]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:87]: why did you remove these C dependencies?

File: recipe/meta.yaml [hunk 6]
```diff
@@ -75,22 +72,30 @@ outputs:
...
         - libprotobuf
```
Comment (hmaarrfk) [inline recipe/meta.yaml:95]:
```suggestion
        - libprotobuf
        - sleef
```

File: recipe/meta.yaml [hunk 7]
```diff
@@ -75,22 +72,30 @@ outputs:
...
-        - sleef
-        - typing
         - libuv
         - pkg-config  # [unix]
-        - typing_extensions
+        - sleef
```
Comment (hmaarrfk) [inline recipe/meta.yaml:98]:
```suggestion
```

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

File: recipe/build_pytorch.sh [hunk 8]
```diff
@@ -159,7 +159,7 @@ else
     fi
     # MKLDNN is an Apache-2.0 licensed library for DNNs and is used
     # for CPU builds. Not to be confused with MKL.
-    export USE_MKLDNN=1
+    # export USE_MKLDNN=1
```
Comment (hmaarrfk) [inline recipe/build_pytorch.sh:162]: i'm somewhat concerned about this change.

File: recipe/build_pytorch.sh [hunk 9]
```diff
@@ -125,7 +124,7 @@ if [[ ${cuda_compiler_version} != "None" ]]; then
     # Even though cudnn is used for CUDA builds, it's good to enable
     # for MKLDNN for CUDA builds when CUDA builds are used on a machine
     # with no NVIDIA GPUs. However compilation fails with mkldnn and cuda enabled.
-    export USE_MKLDNN=OFF
+    # export USE_MKLDNN=OFF
```
Comment (hmaarrfk) [inline recipe/build_pytorch.sh:127]: can we also revert this?

**Comment by hmaarrfk** [comment]
As part of my review, I was looking for:
```
WARNING (pytorch,lib/python3.10/site-packages/torch/lib/libtorch_global_deps.so): Needed DSO lib/libmkl_intel_lp64.so.2 found in ['mkl']
WARNING (pytorch,lib/python3.10/site-packages/torch/lib/libtorch_global_deps.so): .. but ['mkl'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (pytorch,lib/python3.10/site-packages/torch/lib/libtorch_global_deps.so): Needed DSO lib/libmkl_gnu_thread.so.2 found in ['mkl']
WARNING (pytorch,lib/python3.10/site-packages/torch/lib/libtorch_global_deps.so): .. but ['mkl'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (pytorch,lib/python3.10/site-packages/torch/lib/libtorch_global_deps.so): Needed DSO lib/libmkl_core.so.2 found in ['mkl']
```

File: recipe/meta.yaml (hunk 5 above)
Comment (RaulPPelaez) [reply recipe/meta.yaml:87]: I tried to follow this as close as possible https://github.com/pytorch/pytorch#from-source. I figured I could reintroduce anything extra only if needed.

File: recipe/build_pytorch.sh (hunk 8 above)
Comment (RaulPPelaez) [reply recipe/build_pytorch.sh:162]: Ok lets try and set it on again.

File: recipe/meta.yaml (hunk 5 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:87]: these are pretty specific to conda forge omp dependencies. please add them back

File: recipe/meta.yaml [hunk 10]
```diff
@@ -75,22 +72,30 @@ outputs:
...
+        - types-dataclasses
+        - sympy
+        - filelock
+        - networkx
+        - jinja2
+        - fsspec
+        - packaging
+        - optree
         - libcblas              # [blas_impl != "mkl"]
         - liblapack             # [blas_impl != "mkl"]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:94]:
```suggestion
        - liblapack             # [blas_impl != "mkl"]
        - libgomp   # [linux]
        - llvm-openmp    # [osx]
```

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

File: recipe/build_pytorch.sh [hunk 11]
```diff
@@ -56,15 +56,14 @@ for ARG in $CMAKE_ARGS; do
   fi
 done
 unset CMAKE_INSTALL_PREFIX
-export TH_BINARY_BUILD=1
+# export TH_BINARY_BUILD=1
```
Comment (hmaarrfk) [inline recipe/build_pytorch.sh:59]: I understand that openblas is not being found, but i'm unsure why there are some seemingly unrelated changes in the build script.

File: recipe/meta.yaml [hunk 12]
```diff
@@ -31,6 +27,13 @@ build:
   skip: true  # [cuda_compiler_version == "10.2"]
   skip: true  # [cuda_compiler_version == "11.0"]
   skip: true  # [cuda_compiler_version == "11.1"]
+  skip: true  # [py != 310]
+
+requirements:
+  host:
+    - python
+  run:
+    - python
```
Comment (hmaarrfk) [inline recipe/meta.yaml:36]:
```suggestion
```

File: recipe/meta.yaml [hunk 13]
```diff
@@ -40,6 +43,7 @@ outputs:
       detect_binary_files_with_prefix: false
       run_exports:
         - {{ pin_subpackage('pytorch', max_pin='x.x') }}
+      skip: true  # [py != 310]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:46]:
```suggestion
```

File: recipe/meta.yaml [hunk 14]
```diff
@@ -161,8 +177,15 @@ outputs:
...
         - pytorch-cpu                                      # [cuda_compiler_version == "None"]
+      skip: true  # [py != 310]
     requirements:
+      build:
+        - python                                 # [build_platform != target_platform]
+        - cross-python_{{ target_platform }}     # [build_platform != target_platform]
+      host:
+        - python
       run:
+        - python
```
Comment (hmaarrfk) [inline recipe/meta.yaml:188]:
```suggestion
```

File: recipe/meta.yaml [hunk 15]
```diff
@@ -161,8 +177,15 @@ outputs:
       # weigh down cpu implementation and give cuda preference
       track_features:
         - pytorch-cpu                                      # [cuda_compiler_version == "None"]
+      skip: true  # [py != 310]
     requirements:
+      build:
+        - python                                 # [build_platform != target_platform]
+        - cross-python_{{ target_platform }}     # [build_platform != target_platform]
+      host:
+        - python
```
Comment (hmaarrfk) [inline recipe/meta.yaml:186]:
```suggestion
    requirements:
```

**Comment by hmaarrfk** [comment]
sorry for the all the noise. i'm going to try a few things in: https://github.com/conda-forge/pytorch-cpu-feedstock/pull/196

File: recipe/build_pytorch.sh (hunk 11 above)
Comment (RaulPPelaez) [reply recipe/build_pytorch.sh:59]: I just wanted to make sure builds were not failing because of some of these options in the build script. I started by removing them and then add them back when required.

**Comment by hmaarrfk** [comment]
I'm really kot sure what the difference between the two PRs is. I mostly just untabbed everything so I can test locally

**Comment by jakirkham** [comment]
Now that PR ( https://github.com/conda-forge/pytorch-cpu-feedstock/pull/199 ) is in, should we close this?
